File: mac-athlon-agent/AthlonAgent/Infrastructure/Tools/FileWriteTool.swift
import Foundation

struct FileWriteTool: AgentTool {
    let name = "file_write"
    let description = "Create or overwrite a workspace file with backup."
    let parametersSchema: [String: String] = [
        "path": ToolPathDescriptions.workspaceRelativePath,
        "content": "New content"
    ]
    private let guard_: WorkspaceGuard

    init(guard workspaceGuard: WorkspaceGuard) {
        self.guard_ = workspaceGuard
    }

    func invoke(arguments: [String: String]) async throws -> String {
        let path = try ToolArguments.normalizedPath(arguments, tool: name)
        let content = try ToolArguments.required(arguments, name: "content", tool: name)
        let fullPath = try guard_.normalize(path)
        guard guard_.isInsideWorkspace(fullPath) else {
            throw ToolError.outsideWorkspace(fullPath)
        }
        try AtomicFile.writeText(content, to: fullPath)
        let fileName = (fullPath as NSString).lastPathComponent
        return ToolResultFormatter.success("Wrote \(content.count) chars to \(fileName)")
    }
}

File: athlon-work/src/Athlon.Agent.Infrastructure/FileWriteTool.cs
using System.Diagnostics;
using System.Net.Http.Json;
using System.Runtime.Versioning;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using Athlon.Agent.Core;
using Athlon.Agent.Core.Compaction;
using Microsoft.Extensions.DependencyInjection;
using Serilog;
using Serilog.Core;
using Serilog.Events;

namespace Athlon.Agent.Infrastructure;

public sealed class FileWriteTool(WorkspaceGuard guard, AuditLogService audit) : IAgentTool
{
    public ToolDefinition Definition { get; } = new("file_write", "Create or overwrite a workspace file with backup.", new Dictionary<string, string> { ["path"] = ToolPathDescriptions.WorkspaceRelativePath, ["content"] = "New content" }, RequiresApproval: true);

    public async Task<ToolResult> InvokeAsync(ToolInvocation invocation, CancellationToken cancellationToken = default)
    {
        if (!ToolArguments.TryGetNormalizedPath(invocation, out var path, out var error)) return error;
        if (!ToolArguments.TryGetRequired(invocation, "content", out var content, out error)) return error;

        var fullPath = guard.Normalize(path);
        if (!guard.IsInsideWorkspace(fullPath)) return ToolResult.Failure("Outside workspace", fullPath);
        Directory.CreateDirectory(Path.GetDirectoryName(fullPath)!);
        await File.WriteAllTextAsync(fullPath, content, cancellationToken);
        await audit.WriteAsync("file_write", new { path = fullPath, chars = content.Length }, cancellationToken);
        return ToolResult.Success($"Wrote {content.Length} chars to {Path.GetFileName(fullPath)}");
    }
}
